src/platform/posix/posix_file_descriptordial.c

```diff
@@ -0,0 +1,416 @@
+// get an ephemeral port.
+void
+nni_file_descriptor_dial(nni_file_descriptor_dialer *d, const nni_sockaddr *sa, nni_aio *aio)
+{
+	nni_file_descriptor_conn *          c;
+	nni_posix_pfd *         pfd = NULL;
+	struct sockaddr_storage ss;
+	size_t                  sslen;
+	int                     fd;
+	int                     rv;
+	int                     ka;
+	int                     nd;
+
+	if (nni_aio_begin(aio) != 0) {
+		return;
+	}
+
+	if (((sslen = nni_posix_nn2sockaddr(&ss, sa)) == 0) ||
+	    ((ss.ss_family != AF_INET) && (ss.ss_family != AF_INET6))) {
+		nni_aio_finish_error(aio, NNG_EADDRINVAL);
+		return;
+	}
+
+	if ((fd = socket(ss.ss_family, SOCK_STREAM | SOCK_CLOEXEC, 0)) < 0) {
+		nni_aio_finish_error(aio, nni_plat_errno(errno));
+		return;
+	}
+
+	nni_atomic_inc64(&d->ref);
+
+	if ((rv = nni_posix_file_descriptor_alloc(&c, d)) != 0) {
+		nni_aio_finish_error(aio, rv);
+		nni_posix_file_descriptor_dialer_rele(d);
+		return;
+	}
+
+	// This arranges for the fd to be in non-blocking mode, and adds the
+	// poll fd to the list.
+	if ((rv = nni_posix_pfd_init(&pfd, fd)) != 0) {
+		(void) close(fd);
+		// the error label unlocks this
+		nni_mtx_lock(&d->mtx);
+		goto error;
+	}
+
+	nni_posix_file_descriptor_init(c, pfd);
+	nni_posix_pfd_set_cb(pfd, file_descriptor_dialer_cb, c);
+
+	nni_mtx_lock(&d->mtx);
+	if (d->closed) {
+		rv = NNG_ECLOSED;
+		goto error;
+	}
+	if (d->srclen != 0) {
+		if (bind(fd, (void *) &d->src, d->srclen) != 0) {
+			rv = nni_plat_errno(errno);
+			goto error;
+		}
+	}
+	if ((rv = nni_aio_schedule(aio, file_descriptor_dialer_cancel, d)) != 0) {
+		goto error;
+	}
+	if (connect(fd, (void *) &ss, sslen) != 0) {
+		if (errno != EINPROGRESS) {
+			rv = nni_plat_errno(errno);
+			goto error;
+		}
+		// Asynchronous connect.
+		if ((rv = nni_posix_pfd_arm(pfd, NNI_POLL_OUT)) != 0) {
+			goto error;
+		}
+		c->dial_aio = aio;
+		nni_aio_set_prov_data(aio, c);
+		nni_list_append(&d->connq, aio);
+		nni_mtx_unlock(&d->mtx);
+		return;
+	}
+	// Immediate connect, cool!  This probably only happens
+	// on loop back, and probably not on every platform.
+	nni_aio_set_prov_data(aio, NULL);
+	nd = d->nodelay ? 1 : 0;
+	ka = d->keepalive ? 1 : 0;
+	nni_mtx_unlock(&d->mtx);
+	nni_posix_file_descriptor_start(c, nd, ka);
+	nni_aio_set_output(aio, 0, c);
+	nni_aio_finish(aio, 0, 0);
+	return;
+
+error:
+	nni_aio_set_prov_data(aio, NULL);
+	nni_mtx_unlock(&d->mtx);
+	nng_stream_free(&c->stream);
+	nni_aio_finish_error(aio, rv);
+}
+
+static int
+file_descriptor_dialer_set_nodelay(void *arg, const void *buf, size_t sz, nni_type t)
+{
+	nni_file_descriptor_dialer *d = arg;
+	int             rv;
+	bool            b;
+
+	if (((rv = nni_copyin_bool(&b, buf, sz, t)) != 0) || (d == NULL)) {
+		return (rv);
+	}
+	nni_mtx_lock(&d->mtx);
+	d->nodelay = b;
+	nni_mtx_unlock(&d->mtx);
+	return (0);
+}
+
+static int
+file_descriptor_dialer_get_nodelay(void *arg, void *buf, size_t *szp, nni_type t)
+{
```

Comment:
All these can go away... if the user is giving us a TCP socket, then they are expected to have set this up -- the management of such things will be external to NNG for just plain FDs.